packages/components/src/icon/index.tsx

```diff
@@ -0,0 +1,104 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentType, HTMLProps, SVGProps } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	cloneElement,
+	createElement,
+	Component,
+	isValidElement,
+} from '@wordpress/element';
+import { SVG } from '@wordpress/primitives';
+
+/**
+ * Internal dependencies
+ */
+import Dashicon from '../dashicon';
+import type { Icon as DashiconIcon } from '../dashicon/types';
+
+type IconType< P > = DashiconIcon | ComponentType< P > | JSX.Element;
+
+interface BaseProps< P > {
+	/**
+	 * The icon to render. Supported values are: Dashicons (specified as
+	 * strings), functions, WPComponent instances and `null`.
+	 *
+	 * @default null
+	 */
+	icon?: IconType< P >;
+	/**
+	 * The size (width and height) of the icon.
+	 *
+	 * @default `20` (when using Dashicon), `24` otherwise
+	 */
+	size?: number;
+}
+
+type AdditionalProps< T > = T extends ComponentType< infer U >
+	? U
+	: T extends DashiconIcon
+	? SVGProps< SVGSVGElement >
+	: {};
+
+export type Props< P > = BaseProps< P > & AdditionalProps< IconType< P > >;
+
+function Icon< P >( { icon, size, ...additionalProps }: Props< P > ) {
+	if ( 'string' === typeof icon ) {
+		return (
+			<Dashicon
+				icon={ icon }
+				{ ...( additionalProps as HTMLProps< HTMLSpanElement > ) }
+			/>
+		);
+	}
+
+	if ( isValidElement( icon ) && Dashicon === icon.type ) {
+		return cloneElement( icon, {
+			...additionalProps,
+		} );
+	}
+
+	// Icons should be 24x24 by default.
+	const iconSize = size || 24;
```

Comment:
I wonder if would do any harm to specify this as a default arg value instead. Then we could also shorten some of the code by avoiding declaration and referencing an extra variable.

Reply:
Oh yeah, we can do that!